src/variational/advi.jl

```diff
@@ -0,0 +1,105 @@
+"""
+    ADVI(samples_per_step = 10, max_iters = 5000)
+
+Automatic Differentiation Variational Inference (ADVI) for a given model.
+"""
+struct ADVI{AD} <: VariationalInference{AD}
+    samples_per_step # number of samples used to estimate the ELBO in each optimization step
+    max_iters        # maximum number of gradient steps used in optimization
+end
+
+ADVI(args...) = ADVI{ADBackend()}(args...)
+ADVI() = ADVI(10, 5000)
+
+alg_str(::ADVI) = "ADVI"
+
+vi(model::Model, alg::ADVI; optimizer = ADAGrad()) = begin
+    # setup
+    var_info = VarInfo()
+    model(var_info, SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    dists = var_info.dists
+    ranges = var_info.ranges
+    
+    q = MeanField(zeros(num_params), zeros(num_params), dists, ranges)
+
+    # construct objective
+    elbo = ELBO()
+
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    θ = optimize(elbo, alg, q, model; optimizer = optimizer)
+    μ, ω = θ[1:length(q)], θ[length(q) + 1:end]
+
+    # TODO: make mutable instead?
+    MeanField(μ, ω, dists, ranges) 
+end
+
+# TODO: implement optimize like this?
+# (advi::ADVI)(elbo::EBLO, q::MeanField, model::Model) = begin
+# end
+
+function optimize(elbo::ELBO, alg::ADVI, q::MeanField, model::Model; optimizer = ADAGrad())
+    θ = randn(2 * length(q))
+    optimize!(elbo, alg, q, model, θ; optimizer = optimizer)
+
+    return θ
+end
+
+function (elbo::ELBO)(alg::ADVI, q::MeanField, model::Model, θ::AbstractVector{T}, num_samples) where T <: Real
```

Comment:
Instead of a caller version, maybe we should call this function `estimate`?

Reply:
Why?

Reply:
As we are actually estimating it?

Reply:
Sounds reasonable. How does this play with the `objective` function? Do we need both?

Reply:
We could do something like `estimate(elbo::ELBO, ...)` instead of having this `objective`. My concern is that this potentially makes it harder to allow arbitrary functions as objectives since the user then basically has to subtype `VariationalObjective`?

And writing generic `optimize!` becomes more difficult since we now don't know whether or not to directly call the `VariationalObjective` or call `estimate`. This was my original reason for not making some `estimate` function.